metadata/src/main/resources/common/metadata/PartitionRecord.json

```diff
@@ -47,6 +47,13 @@
       "about": "The eligible leader replicas of this partition." },
     { "name": "LastKnownELR", "type": "[]int32", "default": "null", "entityType": "brokerId",
       "versions": "1+", "nullableVersions": "1+", "taggedVersions": "1+", "tag": 2,
-      "about": "The last known eligible leader replicas of this partition." }
+      "about": "The last known eligible leader replicas of this partition." },
+    { "name": "Assignment", "type": "[]ReplicaAssignment", "versions": "2+",
+      "about": "The replica assignment for this partition, sorted by preferred order.", "fields": [
+      { "name": "Broker", "type": "int32", "versions": "2+", "entityType": "brokerId",
+        "about": "The broker ID hosting the replica." },
+      { "name": "Directory", "type": "uuid", "versions": "2+",
+        "about": "The log directory hosting the replica" }
+    ]}
```

Comment:
The `Assignment` field added to `PartitionChangeRecord` is added as a tagged field, yet the `Assignment` field added here in `PartitionRecord` is not.  I think they should both be tagged fields?

Reply:
Why should `PartitionRecord.Assignment` be a tagged field?

All fields in `PartitionChangeRecord` are tagged (except for topic and partition ID). It makes sense that a change record have pretty much all fields be optional, and `Assignment` - like `Replicas` - is no different.

A `PartitionRecord` however, without `Assignment` makes less sense to me. `PartitionRecord.Assignment` should replace `PartitionRecord.Replicas` which is not a tagged field. Any partition should have a set of replicas, now they need a different structure, but the list should still be there.

Reply:
It's a definitely a good shout – some of the assumptions have changed since then. I agree it's attractive to add a new orthogonal dir UUIDs field, that'll be an easier change to handle. And I'm also uncomfortable with not taking any action to deprecate the existing `Replicas` field. 

My concerns are:

* It will be possible to construct a record that is semantically invalid, with a different number of entries in `record.replicas` and `record.logDirs`.
* We'll need extra care and it might be a bit awkward to keep the order of both lists in sync, any failures there will lead to issues that might be hard to detect or diagnose, as there won't be any hard link between both values.
* This problem stems mainly from our recent decision to postpone figuring out how to burn field tags. My understanding is that keeping the Replicas field unchanged is a temporary measure until then. Whereas in the long term every replica should always have log dir. In other words, I don't think we'd consider this if `Replicas` didn't already exist, so it feels like we're sacrificing the end design due to the existing limitation on tagged fields.

Reply:
I would say that this is all accurate, yes.